File: src/day22/Day22.kt
package day22

import java.io.File

const val FILENAME = "src/day22/Input.txt"

val input = File(FILENAME).readText().split("\n\n").map {
    val list = it.split("\n")
    list.subList(1, list.size).map { str -> str.toInt() }.toList()
}.toTypedArray()

fun part01(): Int {
    val player1 = input[0].toMutableList()
    val player2 = input[1].toMutableList()

    while (player1.any() && player2.any()) {
        val p1 = player1.removeAt(0)
        val p2 = player2.removeAt(0)
        if (p1 > p2) {
            player1.add(p1)
            player1.add(p2)
        } else {
            player2.add(p2)
            player2.add(p1)
        }
    }

    return if (player1.any()) winnerScore(player1) else winnerScore(player2)
}

fun winnerScore(deck: List<Int>): Int {
    var winnerScore = 0
    for (i in deck.indices)
        winnerScore += deck[i] * (deck.size - i)
    return winnerScore
}

fun part02(): Int {
    val player1 = input[0].toMutableList()
    val player2 = input[1].toMutableList()

    val p1Win = recursiveCombat(player1, player2)
    return if (p1Win) winnerScore(player1) else winnerScore(player2)
}

fun recursiveCombat(deck1: MutableList<Int>, deck2: MutableList<Int>): Boolean {
    val history = HashSet<Pair<String, String>>()

    while (deck1.any() && deck2.any()) {
        val s1 = deck1.joinToString(",")
        val s2 = deck2.joinToString(",")
        if (history.contains(Pair(s1, s2)))
            return true
        history.add(Pair(s1, s2))

        val p1 = deck1.removeAt(0)
        val p2 = deck2.removeAt(0)
        if (p1 <= deck1.size && p2 <= deck2.size) {
            val p1WinSubgame = recursiveCombat(deck1.take(p1).toMutableList(), deck2.take(p2).toMutableList())
            if (p1WinSubgame) {
                deck1.add(p1)
                deck1.add(p2)
            } else {
                deck2.add(p2)
                deck2.add(p1)
            }
        } else {
            if (p1 > p2) {
                deck1.add(p1)
                deck1.add(p2)
            } else {
                deck2.add(p2)
                deck2.add(p1)
            }
        }
    }

    // If deck1 is empty, player2 wins (returns false). If deck1 has cards, then deck2 must be empty, and player1 wins (returns true)
    return deck1.any()
}

fun main() {
    println("Advent of Code 2020, Day 22")
    println(part01())
    println(part02())
}

File: day22/day22.go
package day22

import (
	"advent-of-code-2020/utils"
	"fmt"
	"strings"
)

const filename = "./day22/input.txt"

var input = func() [][]int {
	strs := strings.Split(utils.ReadString(filename), "\n\n")
	input := make([][]int, len(strs))
	for i, s := range strs {
		input[i] = utils.MapToInt(strings.Split(s, "\n")[1:])
	}
	return input
}()

type pair struct {
	s1, s2 string
}

func part01() int {
	player1 := make([]int, len(input[0]))
	copy(player1, input[0])
	player2 := make([]int, len(input[1]))
	copy(player2, input[1])

	for len(player1) > 0 && len(player2) > 0 {
		p1 := player1[0]
		player1 = player1[1:]
		p2 := player2[0]
		player2 = player2[1:]
		if p1 > p2 {
			player1 = append(player1, p1, p2)
		} else {
			player2 = append(player2, p2, p1)
		}
	}

	if len(player1) > 0 {
		return winnerScore(player1)
	} else {
		return winnerScore(player2)
	}
}

func winnerScore(deck []int) int {
	winnerScore := 0
	for i := range deck {
		winnerScore += deck[i] * (len(deck) - i)
	}
	return winnerScore
}

func part02() int {
	player1 := input[0]
	player2 := input[1]

	p1Win := recursiveCombat(&player1, &player2)
	if p1Win {
		return winnerScore(player1)
	} else {
		return winnerScore(player2)
	}
}

func recursiveCombat(deck1, deck2 *[]int) bool {
	history := utils.NewSet[pair]()

	for len(*deck1) > 0 && len(*deck2) > 0 {
		s1 := strings.Join(utils.MapToStr(*deck1), ",")
		s2 := strings.Join(utils.MapToStr(*deck2), ",")
		if history.Contains(pair{s1, s2}) {
			return true
		}
		history.Add(pair{s1, s2})

		p1 := (*deck1)[0]
		*deck1 = (*deck1)[1:]
		p2 := (*deck2)[0]
		*deck2 = (*deck2)[1:]
		if p1 <= len(*deck1) && p2 <= len(*deck2) {
			deck1Copy := make([]int, p1)
			copy(deck1Copy, (*deck1)[:p1])
			deck2Copy := make([]int, p2)
			copy(deck2Copy, (*deck2)[:p2])
			p1WinSubgame := recursiveCombat(&deck1Copy, &deck2Copy)
			if p1WinSubgame {
				*deck1 = append(*deck1, p1, p2)
			} else {
				*deck2 = append(*deck2, p2, p1)
			}
		} else {
			if p1 > p2 {
				*deck1 = append(*deck1, p1, p2)
			} else {
				*deck2 = append(*deck2, p2, p1)
			}
		}
	}

	// If deck1 is empty, player2 wins (returns false). If deck1 has cards, then deck2 must be empty, and player1 wins (returns true)
	return len(*deck1) > 0
}

func Main() {
	fmt.Println("Advent of Code 2020, Day 22")
	fmt.Println(part01())
	fmt.Println(part02())
}
